Git implements IProjectFilesBrowserService

radektheloner opened the pull request.

radektheloner commented:
So, this is ready to be reviewed now. Among other changes, there are two significant news:
- I've changed the way we mock GitHub. I mean, we still use `nock` but it's been wrapped into a class `GitHubNock` which hopefully makes the code more readable, reduces code duplicity and simplifies writing future tests. I hope you'll like it... :)
- I've reduced `IProjectFilesBrowserService` to only readonly methods. Originally I planned to create a new interface that specifies the other methods but I've changed my mind... I thing that the only place where we would need the other methods will be the reporter. And I believe that we will need `FileSystemService` anyway. So I think there is no need for another interface.

But I can change any of these if you want, of course...

adelkahomolova commented on src/services/git/Git.ts:
```diff
@@ -58,17 +123,50 @@ export class Git {
-  private contentCacheKey(path: string): string {
-    const params = GitHubUrlParser.getOwnerAndRepoName(this.repository.url);
-    return `${params.owner}:${params.repoName}:content:${path}`;
+  private async followSymLinks(path: string, directory?: string): Promise<string> {
```
Do we really want to use this method? See the comment above

radektheloner replied:
ditto

adelkahomolova commented on src/services/git/__MOCKS__/gitHubClientMockFolder/getContributorsStatsResponse.mock.ts:
```diff
@@ -1,3 +1,5 @@
+import { UserItem } from '../../../../../test/helpers/gitHubNock';
```
What about using some alias when the path is that long? But it's just nice to have especially in MOCKS. So maybe ignore this comment :D

radektheloner replied:
Hm, why not. Do we use aliases elsewhere (for my inspiration)?

radektheloner replied:
Hm, why not. @adelkahomolova, do we use aliases elsewhere (for my inspiration)?

adelkahomolova replied:
I don't think so. But I read this article a few days ago. https://dev.to/larswaechter/path-aliases-with-typescript-in-nodejs-4353

prokopsimek replied:
I have tried that and had some errors. ofc it´d be great

radektheloner replied:
I tried my best but still not able to make it work... Not even with the `~` symbol. Can you try that, please? @adelkahomolova, @prokopsimek

prokopsimek replied:
It's not so important. If it doesn't work, leave it as is.

adelkahomolova replied:
There are more important things I have to do on DX scanner. This is nice to have, I'll have it on my mind and try it if I'll have nothing to do :)

adelkahomolova commented on src/services/git/__MOCKS__/gitHubClientMockFolder/getIssueCommentsResponse.mock.ts:
```diff
@@ -1,3 +1,5 @@
+import { UserItem } from '../../../../../test/helpers/gitHubNock';
```
see the comment above

radektheloner replied:
ditto

radektheloner commented on src/scanner/cache/InMemoryCahce.ts:
```diff
@@ -26,7 +26,7 @@ export class InMemoryCache implements ICache {
   async getOrSet<T>(key: string, setter: () => Promise<T>): Promise<T> {
     const previous = this.get(key);
-    if (previous) {
+    if (previous !== undefined) {
```
Because otherwise the case won't be able to store `null`. I mean, if `previous` was `null`, the condition above would be false and the `setter` would be called.

radektheloner commented on src/services/git/Git.ts:
```diff
@@ -19,28 +21,91 @@ export class Git {
-      }
-    });
+  async exists(path: string): Promise<boolean> {
+    const result = await this.getRepoContent(await this.followSymLinks(path));
```
Well, here, the target is GitHub. Symlinks work fine there. Since `FileSystemService.exists` does follow symlinks (though we do not test it anymore) and since GitHub's API doesn't, I have to implement it myself to make it work the same as `FileSystemService`.

radektheloner commented on src/services/git/Git.ts:
```diff
@@ -19,28 +21,91 @@ export class Git {
-      } else {
-        throw ErrorFactory.newInternalError(`${path} is not a file`);
+  async readDirectory(path: string): Promise<string[]> {
+    const result = await this.getRepoContent(await this.followSymLinks(path));
```
ditto

radektheloner commented on src/services/git/Git.ts:
```diff
@@ -19,28 +21,91 @@ export class Git {
+  }
+
+  async readFile(path: string): Promise<string> {
+    const result = await this.getRepoContent(await this.followSymLinks(path));
```
ditto

radektheloner commented on src/services/git/Git.ts:
```diff
@@ -19,28 +21,91 @@ export class Git {
+  }
+
+  async getMetadata(path: string): Promise<Metadata> {
+    let extension: string | undefined = nodePath.posix.extname(path);
```
Here, we talk about GitHub paths that are always POSIX-like.

prokopsimek commented on src/services/git/Git.ts:
```diff
@@ -6,9 +6,11 @@ import { injectable } from 'inversify';
-export class Git {
+export class Git implements IProjectFilesBrowserService {
   private repository: Repository;
   private gitHubClient: GitHubClient;
```
I don't really like that this service is just for GitHub.

radektheloner replied:
So now, `Git` uses a `IContentRepositoryBrowser` instead of `GitHubClient`.

adelkahomolova commented:
Will you change the CONTRIBUTING.md as the GitHubClient no longer exists or should I? :) @radektheloner

radektheloner commented:
Oops... I wlll!

radektheloner commented:
So, I've:
- Merged `GitHubClient` and `GitHubService` and adjusted `CONTRIBUTING.md`
- Moved the caching logic from `Git` to `GitHubService`

@prokopsimek, please, can you please review this?

BTW, I think that `Git` still needs some love... Now, it seems like some kind of a hybrid between a service and inspector.... I think that it should either implement `IFileInspector` and be moved into `inspectors` or be merged with `GitHubService`. I can handle that if you wish...

prokopsimek commented:
:tada: This PR is included in version 1.4.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/dx-scanner)
- [GitHub release](https://github.com/DXHeroes/dx-scanner/releases/tag/v1.4.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
